scripts/inject_commands.py

```diff
@@ -0,0 +1,42 @@
+#!/usr/bin/python3
+
+"""
+Inject nopticon commands into a BMP message stream 
+"""
+
+from argparse import ArgumentParser
+import bmp 
+import nopticon
+
+def main():
+    # Parse arguments
+    arg_parser = ArgumentParser(description='Inject nopticon commands into a BMP message stream')
+    arg_parser.add_argument('-input', dest='input_path', action='store',
+            required=True, help='Path for BMP message stream')
+    arg_parser.add_argument('-output', dest='output_path', action='store',
+            required=True, help='Path for modified message stream')
+    arg_parser.add_argument('-peerchange', dest='peerchange',
+            action='store_true', 
+            help='Print and reset network summary on every peer up/down event')
+    settings = arg_parser.parse_args()
+
+    # Open input and output streams
+    istream = open(settings.input_path, 'r')
+    ostream = open(settings.output_path, 'w')
+
+    # Process input stream
+    for bmp_json in istream.readlines():
+        bmp_msg = bmp.parse_message(bmp_json.strip())
+        if (settings.peerchange 
+                and (bmp_msg.isPeerUp() or bmp_msg.isPeerDown())):
+            ostream.write(nopticon.Command.print_log().json()+'\n')
+            refresh = nopticon.Command.refresh_summary(bmp_msg._timestamp)
```

Comment:
Ordinarily, the timestamp in a BMP message can be a float. But Nopticon's command parser assumes that the timestamp of a command is a uint64_t. Is this being taken into consideration here?

Reply:
GoBGP's peer up/down messages have uint32_t timestamps (i.e., time is measured in seconds). Is Nopticon expecting time in nanoseconds (i.e., a float coerced to a unit64_t)? Also, nanosecond granularity is probably too fine to be trusted. Perhaps we should modify GoBGP to provide a float whose decimal value is in milliseconds.

Reply:
Is the time unit in peer up/down messages different from the ones concerning rule changes? In Nopticon, I treat timestamps as merely uint64s and I do not do any time unit conversions. I currently assume that units are always the same, irrespective of the kind of BMP message that is being sent.

Reply:
The output produced by `gobmpd`  for route monitoring messages has timestamps of the form below (the specific numbers are irrelevant)
```
"Timestamp":1541005934.58444
```
which represents the timestamp in seconds to the precision of 100th of a millisecond (equivalently 10 microseconds).

However the histories produced by Nopticon are of the form below (again, the specific numbers are irrelevant)
```
"history":[1541003188,1541003189,1541003191,1541003328,1541003343,1541003354,1541003371,1541003718,1541003729,1541003744,1541003764,1541004169,1541004184,1541004195,1541004199,1541004199]}
```
which is simply representing seconds to the precision of 1 second.

Something is getting lost [in the parsing](https://github.com/ahorn/nopticon/blob/master/cmd/gobgp_analysis.cc#L443-L460) but I'm not sure what since it seems like rapidjson is automatically truncating the doubles to ints.

Reply:
Route monitoring messages actually have precision to the nanosecond due to [this change](https://github.com/colgate-cs-research/gobgp/commit/72bf376eb457b81ae421104da8210cf7575b939b) I made. (See [Go time package documentation](https://golang.org/pkg/time/#Time.UnixNano) for details.)

Reply:
So it looks like we should change the precision of the "Timestamp" field in Nopticon to a 64-bit float instead of an unsigned 64 bit integer?

Reply:
Floats are a beast. I would prefer if the timestamps are represented as uint64 with whatever unit is desired. The required conversion from floats to uint64 would have to occur here: https://github.com/ahorn/nopticon/blob/master/cmd/gobgp_analysis.cc#L483

Reply:
What unit do you all think will provide the best ordering of events? Is there any way of justifying our choice here besides "10^-n feels right" for some n? 

My gut says n=3 (i.e. milliseconds) btw

Reply:
I [updated GoBGP](https://github.com/colgate-cs-research/gobgp/commit/fed71461877a4f0a58c2a6d58164bf7df5e91b71) so all BMP events have timestamps that are floats, where the integer portion is in seconds and the fractional portion is time in microseconds. I settled on microseconds because this is the finest precision allowed by the [BMP standard](https://tools.ietf.org/html/rfc7854#section-4.2).

I concur with @ericthewry of using millsecond granularity for ordering of events in Nopticon. According to [Wikipedia](https://en.wikipedia.org/wiki/Network_Time_Protocol) (and the references therein): "NTP can usually maintain time to within tens of milliseconds over the public Internet, and can achieve better than one millisecond accuracy in local area networks under ideal conditions."

@ahorn Can Nopticon accept tiimestamps to microsecond precision represented as float64 for the refresh command to be consistent with BMP?

Reply:
Yes, I accept a patch for that.